### PR TITLE
PWGHF: bugfix and updates to the Lc->K0sP workflow

### PR DESCRIPTION
**Pull request by DanielSamitz**
- bugfix in `invMassLcToK0sP` in HfCascCand
- setting B field in candidateCreatorCascade from ccdb (copied this from candidateCreator2(3)Prong, hope it is OK like that)
- updated some histograms in taskLcToK0sP

### CODE REVIEW

File: PWGHF/TableProducer/candidateCreatorCascade.cxx [hunk 1]
```diff
@@ -112,6 +141,14 @@ struct HfCandidateCreatorCascade {
       const auto& trackV0DaughPos = v0.posTrack_as<MyBigTracks>();
       const auto& trackV0DaughNeg = v0.negTrack_as<MyBigTracks>();
 
+      auto collision = bach.collision();
+
+      // set the magnetic field from CCDB
+      auto bc = collision.bc_as<o2::aod::BCsWithTimestamps>();
+      initCCDB(bc, runNumber, ccdb, isRun2 ? ccdbPathGrp : ccdbPathGrpMag, lut, isRun2);
+      bz = o2::base::Propagator::Instance()->getNominalBz(); // z component
```
Comment (lvermunt) [inline PWGHF/TableProducer/candidateCreatorCascade.cxx:149]: Why did you remove `if (runNumber != bc.runNumber()) {  ... }` (see https://github.com/AliceO2Group/O2Physics/blob/8505ec1/PWGHF/TableProducer/candidateCreator2Prong.cxx#L113-L125). 

In addition, I would also keep the comment that is made there on converted data.

Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateCreatorCascade.cxx]: Hi @lvermunt , thanks for your comment. I changed it (I took the chance to also update some histograms again).
Btw, I saw the same also in candidateCreatorOmegac https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateCreatorOmegac.cxx#L107-L110
Maybe one should then update it there too?

Comment (lvermunt) [reply PWGHF/TableProducer/candidateCreatorCascade.cxx]: Hi @DanielSamitz, in the `candidateCreatorOmegac.cxx` it shouldn't be an issue (in my understanding, but please correct me if I'm wrong) due to the iterator (`void process(SelectedCollisions::iterator const& collision, ...`), so there the if-statement would have no effect.

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx [hunk 2]
```diff
@@ -39,14 +39,14 @@ struct HfTaskLcToK0sP {
   HistogramRegistry registry{
     "registry",
     {// data
-     {"hMass", "cascade candidates;inv. mass (#pi K) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0.0f, 5.0f}}}},
+     {"hMass", "cascade candidates;inv. mass (p K_{S}^{0}) (GeV/#it{c}^{2});entries", {HistType::kTH2F, {{500, 1.5f, 3.0f}, {30, 0.0f, 30.0f, "#it{p}_{T} (GeV/#it{c})"}}}},
```
Comment (lvermunt) [inline PWGHF/D2H/Tasks/taskLcToK0sP.cxx:42]: - Your invariant mass binning should still be finer (or just use the range 1.98--2.58 like Lc->pKpi)
- `entries` should be `p_{T}`
- Suggestion to have a configurable pT binning, see https://github.com/AliceO2Group/O2Physics/blob/41eed29/PWGHF/D2H/Tasks/taskLc.cxx#L39 + https://github.com/AliceO2Group/O2Physics/blob/41eed29/PWGHF/D2H/Tasks/taskLc.cxx#L137

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx [hunk 3]
```diff
@@ -39,14 +39,14 @@ struct HfTaskLcToK0sP {
...
-     {"hV0CPA", "cascade candidates;v0 cosine of pointing angle;entries", {HistType::kTH1F, {{110, -0.98f, 1.1f}}}},
+     {"hPtBach", "cascade candidates;bachelor #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH2F, {{100, 0.0f, 10.0f}, {30, 0.0f, 30.0f, "#it{p}_{T} (GeV/#it{c})"}}}},
+     {"hPtV0", "cascade candidates;v0 #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH2F, {{100, 0.0f, 10.0f}, {30, 0.0f, 30.0f, "#it{p}_{T} (GeV/#it{c})"}}}},
+     {"hd0Bach", "cascade candidates;bachelor DCAxy to prim. vertex (cm);entries", {HistType::kTH2F, {{100, -1.0f, 1.0f}, {30, 0.0f, 30.0f, "#it{p}_{T} (GeV/#it{c})"}}}},
+     {"hd0V0pos", "cascade candidates;pos daugh v0 DCAxy to prim. vertex (cm);entries", {HistType::kTH2F, {{200, -5.0f, 5.0f}, {30, 0.0f, 30.0f, "#it{p}_{T} (GeV/#it{c})"}}}},
+     {"hd0V0neg", "cascade candidates;neg daugh v0 DCAxy to prim. vertex (cm);entries", {HistType::kTH2F, {{200, -5.0f, 5.0f}, {30, 0.0f, 30.0f, "#it{p}_{T} (GeV/#it{c})"}}}},
+     {"hV0CPA", "cascade candidates;v0 cosine of pointing angle;entries", {HistType::kTH2F, {{100, 0.98f, 1.0001f}, {30, 0.0f, 30.0f, "#it{p}_{T} (GeV/#it{c})"}}}},
```
Comment (lvermunt) [inline PWGHF/D2H/Tasks/taskLcToK0sP.cxx:49]: - Same comment as above on `entries` and the pT binning for these six histograms
- Unclear why you didn't make `hEta` a 2D histogram

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx [hunk 4]
```diff
@@ -60,6 +60,8 @@ struct HfTaskLcToK0sP {
       registry.add("hPtGenSig", "cascade candidates (MC);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{100, 0., 10.}}});
       registry.add("hCPARecSig", "cascade candidates (matched);cosine of pointing angle;entries", {HistType::kTH1F, {{110, -1.1, 1.1}}});
       registry.add("hCPARecBg", "cascade candidates (unmatched);cosine of pointing angle;entries", {HistType::kTH1F, {{110, -1.1, 1.1}}});
+      registry.add("hV0CPARecSig", "cascade candidates (matched);v0 cosine of pointing angle;entries", {HistType::kTH1F, {{100, 0.98, 1.0001}}});
+      registry.add("hV0CPARecBg", "cascade candidates (unmatched);v0 cosine of pointing angle;entries", {HistType::kTH1F, {{100, 0.98, 1.0001}}});
```
Comment (lvermunt) [inline PWGHF/D2H/Tasks/taskLcToK0sP.cxx:64]: If you make the data histograms 2D, you should also make these MC ones 2D no?

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx [hunk 5]
```diff
@@ -81,14 +83,14 @@ struct HfTaskLcToK0sP {
         continue;
       }
 
-      registry.fill(HIST("hMass"), invMassLcToK0sP(candidate));
+      registry.fill(HIST("hMass"), invMassLcToK0sP(candidate), candidate.pt());
```
Comment (lvermunt) [inline PWGHF/D2H/Tasks/taskLcToK0sP.cxx:86]: Use `auto pt = candidate.pt();` and fill with `pt` instead of `candidate.pt()` (comment holds for all histograms below as well)

File: PWGHF/TableProducer/candidateCreatorCascade.cxx [hunk 6]
```diff
@@ -112,6 +141,22 @@ struct HfCandidateCreatorCascade {
...
+
+      /// Set the magnetic field from ccdb.
+      /// The static instance of the propagator was already modified in the HFTrackIndexSkimCreator,
+      /// but this is not true when running on Run2 data/MC already converted into AO2Ds.
+      auto bc = collision.bc_as<aod::BCsWithTimestamps>();
+      if (runNumber != bc.runNumber()) {
+        LOG(info) << ">>>>>>>>>>>> Current run number: " << runNumber;
+        initCCDB(bc, runNumber, ccdb, isRun2 ? ccdbPathGrp : ccdbPathGrpMag, lut, isRun2);
+        bz = o2::base::Propagator::Instance()->getNominalBz();
+        LOG(info) << ">>>>>>>>>>>> Magnetic field: " << bz;
```
Comment (lvermunt) [inline PWGHF/TableProducer/candidateCreatorCascade.cxx:154]: These `LOG(info)` statements you can leave out I would say, as well as the `// df.print();` below

**Review by lvermunt — COMMENTED** [review]
Hi @DanielSamitz, thanks for addressing my comment. I also have some additional minor comments on your histograms (sorry for not doing it in one go)

File: PWGHF/TableProducer/candidateCreatorCascade.cxx (hunk 1 above)
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateCreatorCascade.cxx]: Ah, I see. That's what got me confused.

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx (hunk 2 above)
Comment (DanielSamitz) [reply PWGHF/D2H/Tasks/taskLcToK0sP.cxx]: OK, I changed that. I tried using `ConfigurableAxis`, but I did not manage to set the default with `hf_cuts_lc_to_k0s_p::vecBinsPt`, so I went for the same option as in https://github.com/AliceO2Group/O2Physics/blob/41eed29/PWGHF/D2H/Tasks/taskLc.cxx#L39 with `Configurable<std::vector<double>>`. I did however not need the extra step as in  https://github.com/AliceO2Group/O2Physics/blob/41eed29/PWGHF/D2H/Tasks/taskLc.cxx#L137    , but could just use `binsPt` directly for the declaration of the axes.

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx (hunk 3 above)
Comment (DanielSamitz) [reply PWGHF/D2H/Tasks/taskLcToK0sP.cxx]: OK, done.

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx (hunk 4 above)
Comment (DanielSamitz) [reply PWGHF/D2H/Tasks/taskLcToK0sP.cxx]: OK, done

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx (hunk 5 above)
Comment (DanielSamitz) [reply PWGHF/D2H/Tasks/taskLcToK0sP.cxx]: OK, done

File: PWGHF/TableProducer/candidateCreatorCascade.cxx (hunk 6 above)
Comment (DanielSamitz) [reply PWGHF/TableProducer/candidateCreatorCascade.cxx]: OK, done

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx [hunk 7]
```diff
@@ -33,40 +33,43 @@ struct HfTaskLcToK0sP {
   Configurable<int> selectionFlagLcToK0sP{"selectionFlagLcToK0sP", 1, "Selection Flag for Lc"};
   Configurable<int> selectionFlagLcbarToK0sP{"selectionFlagLcbarToK0sP", 1, "Selection Flag for Lcbar"};
   Configurable<double> etaCandMax{"etaCandMax", -1., "max. cand. pseudorapidity"};
+  Configurable<std::vector<double>> binsPt{"binsPt", std::vector<double>{hf_cuts_lc_to_k0s_p::vecBinsPt}, "pT bin limits"};
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskLcToK0sP.cxx:36]: Please don't forget to add this in the JSON file in the validation framework.

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx [hunk 8]
```diff
@@ -81,15 +84,16 @@ struct HfTaskLcToK0sP {
...
-      registry.fill(HIST("hMass"), invMassLcToK0sP(candidate));
-      registry.fill(HIST("hPtCand"), candidate.pt());
-      registry.fill(HIST("hPtBach"), candidate.ptProng0());
-      registry.fill(HIST("hPtV0"), candidate.ptProng1());
-      registry.fill(HIST("hd0Bach"), candidate.impactParameter0());
-      registry.fill(HIST("hd0V0pos"), candidate.dcapostopv());
-      registry.fill(HIST("hd0V0neg"), candidate.dcanegtopv());
-      registry.fill(HIST("hV0CPA"), candidate.v0cosPA());
-      registry.fill(HIST("hEta"), candidate.eta());
+      auto candPt = candidate.pt();
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskLcToK0sP.cxx:87]:
```suggestion
      auto ptCand = candidate.pt();
```

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx [hunk 9]
```diff
@@ -105,18 +109,21 @@ struct HfTaskLcToK0sP {
         // Printf("MC Rec.: eta rejection: %g", candidate.eta());
         continue;
       }
+      auto candPt = candidate.pt();
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskLcToK0sP.cxx:112]:
```suggestion
      auto ptCand = candidate.pt();
```

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx [hunk 10]
```diff
@@ -33,40 +33,43 @@ struct HfTaskLcToK0sP {
...
-     {"hEta", "cascade candidates;candidate #it{#eta};entries", {HistType::kTH1F, {{100, -2.0f, 2.0f}}}},
-     {"hSelectionStatus", "cascade candidates;selection status;entries", {HistType::kTH1F, {{5, -0.5f, 4.5f}}}}}};
+  HistogramRegistry registry{"registry"};
 
   void init(InitContext& context)
   {
+    // data
+    registry.add("hMass", "cascade candidates;inv. mass (p K_{S}^{0}) (GeV/#it{c}^{2});p_{T}", {HistType::kTH2F, {{600, 1.98f, 2.58f}, {binsPt, "#it{p}_{T} (GeV/#it{c})"}}});
+    registry.add("hPtCand", "cascade candidates;candidate #it{p}_{T} (GeV/#it{c});entries", {HistType::kTH1F, {{300, 0.0f, 30.0f}}});
+    registry.add("hPtBach", "cascade candidates;bachelor #it{p}_{T} (GeV/#it{c});p_{T}", {HistType::kTH2F, {{100, 0.0f, 10.0f}, {binsPt, "#it{p}_{T} (GeV/#it{c})"}}});
```
Comment (vkucera) [inline PWGHF/D2H/Tasks/taskLcToK0sP.cxx:47]: It's better to mention all the axis variables in the names of multidimensional histograms to make it clear which quantities are filled in and in which order. See the HF naming guidelines.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx (hunk 10 above)
Comment (DanielSamitz) [reply PWGHF/D2H/Tasks/taskLcToK0sP.cxx]: OK, I added a "VsPtCand" for the name of each 2D histogram

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx (hunk 9 above)
Comment (DanielSamitz) [reply PWGHF/D2H/Tasks/taskLcToK0sP.cxx]: OK, done

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx (hunk 8 above)
Comment (DanielSamitz) [reply PWGHF/D2H/Tasks/taskLcToK0sP.cxx]: OK, done

File: PWGHF/D2H/Tasks/taskLcToK0sP.cxx (hunk 7 above)
Comment (DanielSamitz) [reply PWGHF/D2H/Tasks/taskLcToK0sP.cxx:36]: OK, done

**Comment by vkucera** [comment]
Thanks for the fixes
